sdks/v4-sdk/src/hook.ts

```diff
@@ -0,0 +1,77 @@
+  AfterAddLiquidityReturnsDelta = 'afterAddLiquidityReturnsDelta',
+  AfterSwapReturnsDelta = 'afterSwapReturnsDelta',
+  BeforeSwapReturnsDelta = 'beforeSwapReturnsDelta',
+  AfterDonate = 'afterDonate',
+  BeforeDonate = 'beforeDonate',
+  AfterSwap = 'afterSwap',
+  BeforeSwap = 'beforeSwap',
+  AfterRemoveLiquidity = 'afterRemoveLiquidity',
+  BeforeRemoveLiquidity = 'beforeRemoveLiquidity',
+  AfterAddLiquidity = 'afterAddLiquidity',
+  BeforeAddLiquidity = 'beforeAddLiquidity',
+  AfterInitialize = 'afterInitialize',
+  BeforeInitialize = 'beforeInitialize',
+}
+
+export const hookFlagIndex = {
+  [HookOptions.AfterRemoveLiquidityReturnsDelta]: 0,
+  [HookOptions.AfterAddLiquidityReturnsDelta]: 1,
+  [HookOptions.AfterSwapReturnsDelta]: 2,
+  [HookOptions.BeforeSwapReturnsDelta]: 3,
+  [HookOptions.AfterDonate]: 4,
+  [HookOptions.BeforeDonate]: 5,
+  [HookOptions.AfterSwap]: 6,
+  [HookOptions.BeforeSwap]: 7,
+  [HookOptions.AfterRemoveLiquidity]: 8,
+  [HookOptions.BeforeRemoveLiquidity]: 9,
+  [HookOptions.AfterAddLiquidity]: 10,
+  [HookOptions.BeforeAddLiquidity]: 11,
+  [HookOptions.AfterInitialize]: 12,
+  [HookOptions.BeforeInitialize]: 13,
+}
+
+export class Hook {
+  public static permissions(address: string): HookPermissions {
+    address = this._normalizeHookAddress(address)
+
+    return {
+      beforeInitialize: this._hasPermission(address, HookOptions.BeforeInitialize),
+      afterInitialize: this._hasPermission(address, HookOptions.AfterInitialize),
+      beforeAddLiquidity: this._hasPermission(address, HookOptions.BeforeAddLiquidity),
+      afterAddLiquidity: this._hasPermission(address, HookOptions.AfterAddLiquidity),
+      beforeRemoveLiquidity: this._hasPermission(address, HookOptions.BeforeRemoveLiquidity),
+      afterRemoveLiquidity: this._hasPermission(address, HookOptions.AfterRemoveLiquidity),
+      beforeSwap: this._hasPermission(address, HookOptions.BeforeSwap),
+      afterSwap: this._hasPermission(address, HookOptions.AfterSwap),
+      beforeDonate: this._hasPermission(address, HookOptions.BeforeDonate),
+      afterDonate: this._hasPermission(address, HookOptions.AfterDonate),
+      beforeSwapReturnsDelta: this._hasPermission(address, HookOptions.BeforeSwapReturnsDelta),
+      afterSwapReturnsDelta: this._hasPermission(address, HookOptions.AfterSwapReturnsDelta),
+      afterAddLiquidityReturnsDelta: this._hasPermission(address, HookOptions.AfterAddLiquidityReturnsDelta),
+      afterRemoveLiquidityReturnsDelta: this._hasPermission(address, HookOptions.AfterRemoveLiquidityReturnsDelta),
+    }
+  }
+
+  public static hasPermission(address: string, hookOption: HookOptions) {
+    return this._hasPermission(this._normalizeHookAddress(address), hookOption)
+  }
+
+  private static _hasPermission(address: string, hookOption: HookOptions) {
+    // slice only the last 14 bits which are the hook flags and compare with the specific hookOption
```

Comment:
nit but really its slicing the last 16 bits because hex characters are multiples of 4 right?